### ISSUE TITLE
Improve consistency of pedantic lints complaining about badly named types

### ISSUE THREAD

**Issue by AlexWaygood** [astral-sh/ty#3255]
Consider the following file:

```py
from ty_extensions import is_subtype_of
from typing import *

N = NamedTuple("O", [])   # no error
TD = TypedDict("TE", {})  # error: [invalid-argument-type]
T = TypeVar("U")          # error: [invalid-legacy-type-variable]
P = ParamSpec("Q")        # error: [invalid-paramspec]
NT = NewType("N", int)    # error: [invalid-newtype]

reveal_type(N)                            # <class 'O'>
reveal_type(is_subtype_of(N, tuple[()]))  # ConstraintSet[Literal[True]]

reveal_type(TD)                                       # <class 'TE'>
reveal_type(is_subtype_of(TD, Mapping[str, object]))  # ConstraintSet[Literal[True]]

reveal_type(NT)                      # Unknown
reveal_type(is_subtype_of(NT, int))  # ConstraintSet[Literal[False]]

# error: [invalid-argument-type] "`TypeVar` is not a valid argument to `Generic`"
# error: [invalid-type-form] "Variable of type `ParamSpec` is not allowed in a type expression"
class X(Generic[T, P]):
    x: T
    # error: [invalid-type-form] "The first argument to `Callable` must be either a list of types, ParamSpec, Concatenate, or `...`"
    y: Callable[P, int]
```

There are a number of inconsistencies here and it would be nice to clean them up:
- We don't complain about the `NamedTuple` class being assigned to a different name than the string passed in as the first argument. We probably should, for consistency with the other cases.
- This is really the same error across all cases, but we currently use four different error codes for it. I think we should add a new, dedicated error code, use the same error code for all cases, and make it warning-level by default. This is a pedantic lint rather than something that's necessary to achieve soundness.
- Our error recovery for the `NamedTuple` and `TypedDict` cases is far better than it is for the `NewType`, `TypeVar` and `ParamSpec` cases. for `N = NamedTuple("O", [])`, we are still able to infer that `N` is a `NamedTuple` class with 0 fields; for `TD = TypedDict("TE", {})`, we are still able to infer that `TD` is a `TypedDict` class with 0 items. For the other three types, our error recovery is currently quite bad, and this can lead to lots of cascading errors if you accidentally make a typo and pass the wrong string as the first argument to a `TypeVar` constructor call, for example. (I've had this experience locally -- it was confusing!) I think we should be able to have error recovery for `TypeVar`, `ParamSpec` and `NewType` that's just as good as the `TypedDict`/`NamedTuple` cases: we should be able to emit a single diagnostic at the point where the type is constructed, but then gracefully recover from it rather than inferring a worse type as a result of the constructor call.

The diagnostic range is also currently inconsistent. For the `TypeVar`, `ParamSpec` and `NewType` cases, the diagnostic range covers the left-hand side of the assignment. For the `TypedDict` case, the diagnostic range covers the first argument passed to the constructor. I think the `TypedDict` diagnostic range is better here:

<img width="1428" height="338" alt="Image" src="https://github.com/user-attachments/assets/f5867080-4aee-456e-a910-021f2f4caba0" />

The above screenshot also demonstrates some inconsistencies in syntax highlighting between the different kinds of constructor calls. I'm not sure if that might be naturally fixed by fixing some of the above issues or not.

**Comment by AlexWaygood** [comment]
@charliermarsh, this feels related to your work on functional classes -- don't know if you'd be interested in working on this!

**Comment by charliermarsh** [comment]
Always interested.
